File: app/src/main/java/com/example/had/SignupActivity.java
package com.example.had;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;


import java.util.HashMap;

public class SignupActivity extends AppCompatActivity {

    private static final String TAG = "Register";
    private EditText mEmailText, mPasswordText, mPasswordcheckText, mNameText, mIdText,  mBirthText;
    private Button mRegisterBtn, mPasswordBtn;
    private RadioGroup mRadioGroup;
    private String mSex;

    private FirebaseAuth firebaseAuth;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);

        firebaseAuth =  FirebaseAuth.getInstance();

        mIdText = findViewById(R.id.id);
        mEmailText = findViewById(R.id.email);
        mPasswordText = findViewById(R.id.pw);
        mPasswordcheckText = findViewById(R.id.pwcheck);
        mRegisterBtn = findViewById(R.id.signupbutton);
        mNameText = findViewById(R.id.name);
        mBirthText = findViewById(R.id.birth);
        mPasswordBtn = findViewById(R.id.pwbtn);
        mRadioGroup = findViewById(R.id.radiogroup);

        mRadioGroup.setOnCheckedChangeListener(radioGroupButtonChangeListener);

        mPasswordBtn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {
                String pwd = mPasswordText.getText().toString().trim();
                String pwdcheck = mPasswordcheckText.getText().toString().trim();

                if(pwd.equals(pwdcheck)) {
                    Toast.makeText(SignupActivity.this, "비밀번호가 일치합니다.", Toast.LENGTH_SHORT).show();
                } else{
                    Toast.makeText(SignupActivity.this, "비밀번호가 틀렸습니다. 다시 입력해 주세요.", Toast.LENGTH_SHORT).show();
                }
            }
        });


        mRegisterBtn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {

                //가입 정보 가져오기//
                String email = mEmailText.getText().toString().trim();
                //String uid = mIdText.getText().toString().trim();
                String pwd = mPasswordText.getText().toString().trim();
                String pwdcheck = mPasswordcheckText.getText().toString().trim();


                if(pwd.equals(pwdcheck)) {
                    Log.d(TAG, "등록 버튼 " + email + " , " + pwd);
                    final ProgressDialog mDialog = new ProgressDialog(SignupActivity.this);
                    mDialog.setMessage("가입중입니다...");
                    mDialog.show();


                    //파이어베이스에 신규계정 등록하기
                    firebaseAuth.createUserWithEmailAndPassword(email, pwd).addOnCompleteListener(SignupActivity.this, new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull Task<AuthResult> task) {
                            //가입 성공시
                            if (task.isSuccessful()) {
                                mDialog.dismiss();

                                FirebaseUser user = firebaseAuth.getCurrentUser();
                                String email = user.getEmail();
                                String uid = mIdText.getText().toString().trim();
                                String name = mNameText.getText().toString().trim();
                                String birth = mBirthText.getText().toString().trim();
                                String sex = mSex;

                                //해쉬맵 테이블을 파이어베이스 데이터베이스에 저장
                                HashMap<Object, String> hashMap = new HashMap<>();

                                hashMap.put("uid", uid);
                                hashMap.put("email", email);
                                hashMap.put("name", name);
                                hashMap.put("birth", birth);
                                hashMap.put("sex", sex);


                                FirebaseDatabase database = FirebaseDatabase.getInstance();
                                DatabaseReference reference = database.getReference("Users");
                                reference.child(uid).setValue(hashMap);


                                //가입이 이루어져을시 가입 화면을 빠져나감.
                                Intent intent = new Intent(SignupActivity.this, LoginActivity.class);
                                startActivity(intent);
                                finish();
                                Toast.makeText(SignupActivity.this, "회원가입에 성공하셨습니다.", Toast.LENGTH_SHORT).show();
                            } else {
                                mDialog.dismiss();
                                Toast.makeText(SignupActivity.this, "회원가입에 실패하셨습니다.", Toast.LENGTH_SHORT).show();
                                return;  //해당 메소드 진행을 멈추고 빠져나감.
                            }
                        }
                    });

                    //비밀번호 오류시
                }else{

                    Toast.makeText(SignupActivity.this, "비밀번호가 틀렸습니다. 다시 입력해 주세요.", Toast.LENGTH_SHORT).show();
                    return;
                }
            }
        });

    }
    RadioGroup.OnCheckedChangeListener radioGroupButtonChangeListener=new RadioGroup.OnCheckedChangeListener() {
        @Override
        public void onCheckedChanged(RadioGroup group, int checkedId) {
            if(checkedId == R.id.male_rbt){
                mSex = "male";
            } else{
                mSex = "female";
            }
        }
    };

}

File: app/src/main/java/com/example/had/SignupActivity.kt
package com.example.had

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.had.databinding.ActivitySignupBinding
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.ktx.auth
import com.google.firebase.database.ktx.database
import com.google.firebase.ktx.Firebase


class SignupActivity : AppCompatActivity() {
    private val TAG = "SignUp"
    private lateinit var auth: FirebaseAuth
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivitySignupBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val sAdapter = ArrayAdapter.createFromResource(this, R.array.catrgories, android.R.layout.simple_spinner_dropdown_item)
        binding.categoryComboBox.setAdapter(sAdapter);
        // Initialize Firebase Auth
        auth = Firebase.auth
        val database = Firebase.database
        val reference = database.getReference("Users")

        binding.pwbtn.setOnClickListener {
            if (binding.pw.text.toString() != binding.pwcheck.text.toString()) {
                binding.textView14.setVisibility(View.VISIBLE)
                binding.textView.setVisibility(View.INVISIBLE)
            }
            else {
                binding.textView14.setVisibility(View.INVISIBLE)
                binding.textView.setVisibility(View.VISIBLE)
            }
            /*if (binding.pw.text.equals(binding.pwcheck.text))
                binding.textView14.setVisibility(View.INVISIBLE)
            else
                binding.textView14.setVisibility(View.VISIBLE)*/
            Log.d("pw : ", binding.pw.text.toString())
            Log.d("pwcheck : ", binding.pwcheck.text.toString())

        }

        binding.signupbutton.setOnClickListener {

            var name = binding.name.text.toString().trim()
            var pw = binding.pw.text.toString().trim()
            var email = binding.email.text.toString().trim()
            var emailad = binding.categoryComboBox.toString().trim() //db 추가해주세요
            email.plus(emailad)
            var birth = binding.birth.text.toString().trim()
            var sex : String = ""
            var uid : String = ""
            when (binding.radiogroup.checkedRadioButtonId) {
                R.id.male_rbt -> sex = "male".trim()
                R.id.female_rbt -> sex = "female".trim()
            }

            auth.createUserWithEmailAndPassword(email, pw).addOnCompleteListener(this) { task ->
                if (task.isSuccessful) {
                    // Sign in success, update UI with the signed-in user's information
                    Log.d(TAG, "createUserWithEmail:success")

                    val user = auth.currentUser
                    val hashMap: HashMap<Any, String> = HashMap()
                    user?.let {
                        uid = user.uid
                    }

                    hashMap.put("email", email)
                    hashMap.put("name", name)
                    hashMap.put("birth", birth)
                    hashMap.put("sex", sex)

                    reference.child(uid).setValue(hashMap)

                    Toast.makeText(baseContext, "회원가입에 성공하셨습니다.", Toast.LENGTH_SHORT).show()

                    val intent = Intent(this@SignupActivity, LoginActivity::class.java)
                    startActivity(intent)
                    finish()
                } else {
                    // If sign in fails, display a message to the user.
                    Log.w(TAG, "createUserWithEmail:failure", task.exception)
                    Toast.makeText(baseContext, "이미 등록된 이메일입니다.", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }
}